src/tracked_struct.rs

```diff
@@ -635,20 +638,26 @@ where
             unsafe { (*data).assume_init_mut() }.take_memo_table()
         });
 
+        self.clear_memos(zalsa, &mut memo_table, id);
+    }
+
+    /// Clears the given memo table.
+    pub(crate) fn clear_memos(&self, zalsa: &Zalsa, memo_table: &mut MemoTable, id: Id) {
         // SAFETY: We use the correct types table.
-        let table = unsafe { self.memo_table_types.attach_memos_mut(&mut memo_table) };
+        let table = unsafe { self.memo_table_types.attach_memos_mut(memo_table) };
+
         // `Database::salsa_event` is a user supplied callback which may panic
         // in that case we need a drop guard to free the memo table
         struct TableDropGuard<'a>(MemoTableWithTypesMut<'a>);
         impl Drop for TableDropGuard<'_> {
             fn drop(&mut self) {
-                // SAFETY: We have verified that no more references to these memos exist and so we are good
+                // SAFETY: We have `&mut MemoTable`, so no more references to these memos exist and we are good
                 // to drop them.
                 unsafe { self.0.drop() };
             }
         }
         let mut table_guard = TableDropGuard(table);
-        // SAFETY: We have verified that no more references to these memos exist and so we are good
+        // SAFETY: We have `&mut MemoTable`, so no more references to these memos exist and we are good
```

Comment:
I think the `self.free_list.push` shouldn't be in `clear_memos` because we don't want to push the memo into the free list when updating a tracked struct

Reply:
Good catch, thanks.